climada/engine/supplychain.py

```diff
@@ -0,0 +1,395 @@
+
+            # if folder does not exist -> create folder and download file
+            except FileNotFoundError:
+                os.chdir(SYSTEM_DIR)
+                download_link = WIOD_FILE_LINK + file_name
+                download_file(download_link)
+                LOGGER.debug('Downloading WIOT table for year %s', year)
+
+                mriot = pd.read_excel(os.path.join(results_folder, file_name),
+                                  engine='pyxlsb')
+        else:
+            mriot = pd.read_excel(os.path.join(file_path, file_name),
+                                  engine='pyxlsb')
+
+        row_st,row_end=rows_range
+        col_sect,row_sect_end=cols_sect_range
+        col_data_st,col_data_end=cols_data_range
+
+        sectors = mriot.iloc[row_st:row_sect_end, col_sect].values
+        countries_iso3 = np.unique(mriot.iloc[row_st:row_end, col_iso3])
+        # move Rest Of World (ROW) at the end of the array as countries
+        # are in chronological order
+        idx_row = np.where(countries_iso3 == 'ROW')[0][0]
+        countries_iso3 = np.hstack([countries_iso3[:idx_row],
+                                    countries_iso3[idx_row+1:], np.array('ROW')])
+
+        mriot_data = mriot.iloc[row_st:row_end,
+                                col_data_st:col_data_end].values
+        total_prod = mriot.iloc[row_st:row_end, -1].values
+
+        n_countries = len(countries_iso3)
+        n_sectors = len(sectors)
+
+        cntry_pos = {}
+        countries = []
+        for i, _ in enumerate(countries_iso3):
+            iso3 = countries_iso3[i]
+            cntry_pos.update({iso3: range(i*n_sectors, n_sectors*(i+1))})
+
+            try:
+                countries.append(countries_by_alpha3[iso3][0])
+            except KeyError:
+                countries.append('Rest of World')
+
+        self.mriot_data = mriot_data
+        self.countries = np.array(countries, dtype=str)
+        self.countries_iso3 = countries_iso3
+        self.sectors = sectors
+        self.total_prod = total_prod
+        self.n_countries = n_countries
+        self.n_sectors = n_sectors
+        self.cntry_pos = cntry_pos
+        self.mriot_type = 'wiod'
+
+    def calc_sector_direct_impact(self, hazard, exposure, imp_fun_set,
+                                  sec_subsec=None, sector_type=None):
+        """Calculate for each country/sector-combination the direct impact per year.
+        I.e. compute one year impact set for each country/sector combination. Returns
+        the notion of a supplychain year impact set, which is a dataframe with size
+        (n years) * ((n countries)*(n sectors)).
+
+        Parameters:
+            hazard (Hazard): Hazard object for impact calculation.
+            exposure (Exposures): Exposures object for impact calculation.
+            imp_fun_set (ImpactFuncSet): Set of impact functions.
+            sec_subsec (list): User-defined list with the start and end positions
```

Comment:
Dealing with row and column names is not possible. We are talking about over 2400 rows/columns, where names in fact replicate. For n sectors, you haven n rows/cols for country 1, then n rows/cols for country 2,...., n rows/cols for country m.. and they are named the same. The easiest way to access it is to know at which pos is your country, and then which of the n sectors you need. Which also points to your comment below about this piece of code:

`            values_pos_cntry = range(step_in_table+init_pos,
                                     step_in_table+end_pos)
            subsec_cntry_prod = self.mriot_data[values_pos_cntry].sum(axis=1)
`

this is explained in the tutorial

Reply:
Yes. But why then not use the already existing separator `self.cntry_pos` instead of building them new every time? Also, why define the `subsec` as neigbours in the table?

Here I would expect that you use two functions to select
1. `self.cntry_pos` to select by country
2. subsec_pos to select the position of the sections inside of the country rows/columns. This should simply be a list of the positions instead of the start/end positions, as this restricts the use without clear reason.

This way the code becomes clearer and it is more robust to errors in indexing.

Reply:
The code has been updated.. now I use `self.cntry_pos` and allow non continuous indexing